File: leetcode/34.java
class Solution {
    public int[] searchRange(int[] arr, int target) {
        int[] result = new int[2];
        result[0]=lb(arr, target, 0, arr.length);
        result[1]=rb(arr, target, 0, arr.length);
        return result;
    }

    private int lb(int[] arr, int target, int left, int right){
        // empty
        if (left >= right) return -1;
        // 1 element
        if (left == right - 1)
            if (arr[left] == target) return left;
            else return -1;
        // 2 elements
        if (left == right - 2)
            if (arr[left] == target) return left;
            else if (arr[left+1] == target) return left+1;
            else return -1;

        // more than 3
        int mid = (left + right)/2;
        if (arr[mid] == target) return lb(arr, target, left, mid+1);
        else if (arr[mid] < target) return lb(arr,target,mid+1, right);
        else return lb(arr,target,left,mid);
    }


    private int rb(int[] arr, int target, int left, int right){
        // empty
        if (left >= right) return -1;
        // 1 element
        if (left == right - 1)
            if (arr[left] == target) return left;
            else return -1;
        // 2 elements
        if (left == right - 2)
            if (arr[left+1] == target) return left+1;
            else if (arr[left] == target) return left;
            else return -1;

        // more than 3
        int mid = (left + right)/2;
        if (arr[mid] == target) return rb(arr, target, mid, right);
        else if (arr[mid] < target) return rb(arr,target,mid+1, right);
        else return rb(arr,target,left,mid);
    }
}

File: leetcode/34.cpp
template<typename T>
int binary_search(
        const std::vector<T>& vec, 
        int start, 
        int end, 
        const T& key,
        bool lower_bound)
{
    // end condition
    if(start >= end) return (lower_bound?vec.size():-1); 

    // mid
    int middle = start + ((end - start) / 2);
    // std::cout << start << " " << end <<" " << middle << std::endl;

    if(vec[middle] == key) { 
        if (lower_bound)  
            return min(middle, binary_search(vec, start, middle, key, lower_bound));
        else
            return max(middle, binary_search(vec, middle + 1, end, key, lower_bound));
    }
    else if(vec[middle] > key) {
        return binary_search(vec, start, middle, key, lower_bound);
    }
    else {
        return binary_search(vec, middle + 1, end, key, lower_bound);
    }
}


class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        
        int lb = binary_search(nums, 0, (int)nums.size(), target, true);
        
        vector<int> result;
        if (lb < nums.size()){
            result.push_back(lb);
            result.push_back(binary_search(nums,0,nums.size(),target,false));
        }
        else {
            result.push_back(-1);
            result.push_back(-1);
        }
        return result;
    }
    
};



